tests/middleware_exceptions/middleware.py

```diff
@@ -66,3 +67,21 @@ def process_template_response(self, request, response):
 class NotFoundMiddleware(BaseMiddleware):
     def __call__(self, request):
         raise Http404('not found')
+
+
+class TeapotMiddleware(BaseMiddleware):
+    def __call__(self, request):
+        response = self.get_response(request)
+        response.status_code = 418
+        return response
+
+
+@async_middleware
+def async_teapot_middleware(get_response):
+
+    async def middleware(request):
+        response = await get_response(request)
+        response.status_code = 418
+        return response
+
+    return middleware
```

Comment:
Would it work to declare the async middleware like this?
```
class AsyncBaseMiddleware(BaseMiddleware):
    _is_async = True

class AsyncTeapotMiddleware(AsyncBaseMiddleware):
    async def __call__(self, request):
        response = await self.get_response(request)
        response.status_code = 418
        return response
```

Reply:
@nediamond I could be missing something but I don't Python offers a way to define an `async` `__call__` method.

In this case `AsyncTeapotMiddleware()()` returns a `coroutine` but `inspect.iscoroutinefunction(AsyncTeapotMiddleware())` returns `False`.

Maybe we could use `inspect.iscoroutinefunction(getattr(Foo, '__call__', None))` to avoid the `_is_async` annotation?

Reply:
the following with an async `__call__` works for me!
```
class Test:
	async def __call__(self):
		return True
```
and
`inspect.iscoroutinefunction(Test.__call__) -> True`

Reply:
That's what `inspect.iscoroutinefunction(getattr(Foo, '__call__', None))` does above.

What I mean is that it's probably an abuse of Python data model. For example, 

```python
def Test:
    async def __iter__(self):
        pass

assert inspect.iscoroutinefunction(Test.__iter__)
```

Won't fail but `__aiter__` should be used for this purpose. There's no analogous `__acall__` for `__call__` and it's not clear to me whether `async def __call__` is an abuse or not.

Reply:
Makes sense, looks like this is the relevant PEP section:
https://www.python.org/dev/peps/pep-0492/#why-magic-methods-start-with-a

not sure its an issue here since async `__call__` usage doesn't look to have keyword dependencies

Reply:
The initial comment is how async middleware works in this patch anyway - you `async def` the call, we detect it, and run it async. I'm not sure what change you're asking for?

Reply:
Sorry, wasn't trying to request a change, was thinking about how async middleware would be written and just seeking clarification.

Reply:
Ah, no problem. The test middleware is maybe not the best example, I think once we convert a few of the built-in Django middlewares it'll be more obvious how you can write them.